core/store/src/trie/from_flat.rs

```diff
@@ -0,0 +1,61 @@
+use crate::flat::{store_helper, FlatStorageManager};
+use crate::{ShardTries, Store, Trie, TrieConfig, TrieDBStorage, TrieStorage};
+use near_primitives::{shard_layout::ShardUId, state::FlatStateValue};
+use std::time::Instant;
+
+pub fn construct_trie_from_flat(store: Store, write_store: Store, shard_uid: ShardUId) {
+    let timer = Instant::now();
+    let trie_storage = TrieDBStorage::new(store.clone(), shard_uid);
+
+    let mut entries =
+        store_helper::iter_flat_state_entries(shard_uid, &store, None, None).map(|entry| {
+            let (key, value) = entry.unwrap();
+            let value = match value {
+                FlatStateValue::Ref(ref_value) => {
+                    trie_storage.retrieve_raw_bytes(&ref_value.hash).unwrap().to_vec()
+                }
+                FlatStateValue::Inlined(inline_value) => inline_value,
+            };
+            (key, value)
+        });
+
+    // new ShardTries for write storage location
+    let tries = ShardTries::new(
+        write_store.clone(),
+        TrieConfig::default(),
+        &[shard_uid],
+        FlatStorageManager::new(write_store),
+    );
+    let mut trie_root = Trie::EMPTY_ROOT;
+
+    while let Some(batch) = get_batch(&mut entries) {
+        // Apply and commit changes
+        let batch_size = batch.len();
+        let new_trie = tries.get_trie_for_shard(shard_uid, trie_root);
+        let trie_changes = new_trie.update(batch).unwrap();
+        let mut store_update = tries.store_update();
+        tries.apply_all(&trie_changes, shard_uid, &mut store_update);
```

Comment:
I was pretty surprised by this - all you are doing seems to be writing the kv pairs directly to the trie, and yet that is efficient enough that we don't need to do anything special! I guess the trie already implements batching correctly? Or am I missing something?

Reply:
Behind the scenes, from my understanding, we first update all the states in memory and then commit the changes to store, which makes writing new entries efficient I think.